src/Http/Controllers/UserController.php

```diff
@@ -70,53 +67,33 @@ public function register(Request $request)
             $user->roles($roles);
         }
 
-        // TODO: Registering event
+        if (UserRegistering::dispatch($user) === false) {
+            return $this->userRegistrationSuccess(true);
+        }
 
         $user->save();
 
         UserRegistered::dispatch($user);
 
         Auth::login($user);
 
-        $response = $request->has('_redirect') ? redirect($request->get('_redirect')) : back();
-
-        session()->flash('user.register.success', __('Registration successful.'));
-
-        return $response;
+        return $this->userRegistrationSuccess();
     }
 
-    /**
-     * Run the `user.registering` event.
-     *
-     * This allows the registration to be short-circuited before it gets saved and show errors.
-     * Or, the user may be modified. Lastly, an addon could just 'do something' here without
-     * modifying/stopping the registration.
-     *
-     * Expects an array of event responses (multiple listeners can listen for the same event).
-     * Each response in the array should be another array with an `errors` array.
-     *
-     * @param  UserContract $user
-     * @return MessageBag
-     */
-    protected function runRegisteringEvent(UserContract $user)
+    private function userRegistrationFailure($errors = null)
     {
-        $errors = [];
+        $errorResponse = request()->has('_error_redirect') ? redirect(request()->input('_error_redirect')) : back();
 
-        $responses = event('user.registering', $user);
+        return $errorResponse->withErrors($errors, 'user.register');
+    }
 
-        foreach ($responses as $response) {
-            // Ignore any non-arrays
-            if (! is_array($response)) {
-                continue;
-            }
+    private function userRegistrationSuccess(bool $silentFailure = false)
+    {
+        $response = request()->has('_redirect') ? redirect(request()->get('_redirect')) : back();
 
-            // If the event returned errors, tack those onto the array.
-            if ($response_errors = array_get($response, 'errors')) {
-                $errors = array_merge($response_errors, $errors);
-                continue;
-            }
-        }
+        session()->flash('user.register.success', __('Registration successful.'));
+        session()->flash('user.registered', ! $silentFailure);
```

Comment:
We should name this `user.register.user_created` and add an appropriate variable to the tag.

That lives in this trait. I don't love how it's got form-specific stuff in here, but we can leave that refactor for another time. For now, you can duplicate this block and change it to `user_created`.

https://github.com/statamic/cms/blob/a3505b787c003cdd3e22db37013305a93391ac4c/src/Tags/Concerns/GetsFormSession.php#L23-L28